### DISCUSSION TITLE
[firebase_messaging] How to handle background push messages without never opening app (e.g. after app update)?

### DISCUSSION THREAD

**Discussion by timukasr** [firebase/flutterfire#9359] [Q&A] [answered]
We are rewriting existing app over to Flutter and use firebase_messaging to handle push messages on Android. Everything is working fine, but we have issue when old app is updated to Flutter app in Play Store: push messages do not work until user manually opens the app. In our app, after initial setup, users only interact with the app when push messages arrive so hoping/waiting for them to manually open the app to fix the push messages is bad user experience.

Are there any solutions or suggestions how to best fix this. E.g. a way to register default background handler in Android side without executing Flutter UI?

**Comment by timukasr** [comment, marked as answer]
Came up with following solution:
On Android side, registered `android.intent.action.MY_PACKAGE_REPLACED` BroadcastReceiver, which is called after app is updated. This creates new FlutterEngine:
```kotlin
val engine = FlutterEngine(context)
engine.dartExecutor.executeDartEntrypoint(
    DartExecutor.DartEntrypoint(
        FlutterInjector.instance().flutterLoader().findAppBundlePath(),
        "packageReplacedReceiver"
    )
)
```
Next registered the method on Flutter side in `main.dart`:
```dart
@pragma("vm:entry-point")
void packageReplacedReceiver() {
  FirebaseMessaging.onBackgroundMessage(_firebaseMessagingBackgroundHandler);
}
```
So when app is updated, background handler is registered and when first push message arrives app behaves correctly.

**Reply by irjayjay** [reply]
Hi. I've been trying your solution for most of today. I can't get it to work yet.
Some questions:

1. Where exactly did you put the packageReplacedReceiver() method? just directly into the main.dart? Not in any class or anything? You mention registering it, but I don't see how it's registered.
2. I placed your top snippet inside a BroadcastReceiver's onReceive() method. Is that correct?
3. How do I debug whether the FirebaseMessaging.onBackgroundMessage() actually runs? Breakpoints don't work.

Any help would be greatly appreciated!

**Reply by timukasr** [reply]
> Where exactly did you put the packageReplacedReceiver() method? just directly into the main.dart? Not in any class or anything? You mention registering it, but I don't see how it's registered.

Added following to `AndroidManifest.xml` under `<application>`:
```xml
<receiver android:name=".PackagedReplacedReceiver" android:enabled="true" android:exported="true" >
	<intent-filter>
		<action android:name="android.intent.action.MY_PACKAGE_REPLACED"/>
		<category android:name="android.intent.category.DEFAULT" />
	</intent-filter>
</receiver>
```
And then created the receiver class `PackagedReplacedReceiver.kt`:
```kotlin
package com.example

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import io.flutter.FlutterInjector
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.dart.DartExecutor
import io.flutter.plugin.common.MethodChannel

class PackagedReplacedReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context?, intent: Intent?) {
        val action = intent?.action

        if (context == null || action != Intent.ACTION_MY_PACKAGE_REPLACED) {
            return
        }

        val engine = FlutterEngine(context)

        val channel = MethodChannel(
            engine.dartExecutor.binaryMessenger,
            "com.example/packageReplacedReceiver",
        )
        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                "end" -> {
                    result.success(null)
                    engine.destroy()
                }
                else -> {
                    result.notImplemented()
                }
            }
        }

        engine.dartExecutor.executeDartEntrypoint(
            DartExecutor.DartEntrypoint(
                FlutterInjector.instance().flutterLoader().findAppBundlePath(),
                "packageReplacedReceiver",
            ),
        )
    }
}
```
Not sure, if the method channel part is actually needed.

Testing is harder. Probably should just log something and check with logcat.

**Reply by irjayjay** [reply]
Hi. 
Thank you for your reply. Another member of my team figured it out yesterday. 
Thank you so much for your solution, now our users that haven't reopened our app will get notifications again.

Some notes:
1. The Flutter method should be outside of the main() method on the main.dart - for future reference if anyone else stumbles upon this.
2. Yes, can confirm, the MethodChannel code is unnecessary.

**Reply by irjayjay** [reply]
Hi again,
It seems we now have another issue. 
Our app runs fine in debug mode, but once you build a release, it complains about missing methods.
We've added the `@pragma("vm:entry-point")` tags, but it seems tree shaking is still occurring.

Did you run into the same issues by any chance?

`Dart Unhandled Exception: NoSuchMethodError: No top-level getter 'packageReplacedReceiver' declared.`
